File: 2021/day12/main.py
from collections import defaultdict, Counter


with open("input.txt") as f:
    lines = f.read().splitlines()

root_mapping = defaultdict(list)
for line in lines:
    a, b = line.split('-')
    if b != "start" and a != "end":
        root_mapping[a].append(b)
    if a != "start" and b != "end":
        root_mapping[b].append(a)

paths = []


def find_paths(path, node):
    if node not in ["start", "end"] and node.islower():
        # Part 1
        # if node in path:
        #     return []
        # Part 2
        count = Counter([x for x in path if x.islower()])
        if any((v > 1 for v in count.values())) and count[node] >= 1:
            return []
    if node == "start" and len(path) > 0:
        return []
    path.append(node)
    if node == "end":
        paths.append(path)
        return []

    children = list(filter(None, (find_paths(path.copy(), child) for child in root_mapping[node])))
    return children


find_paths([], "start")

print(len(paths))

File: 2021/day12/src/main.rs
use std::fs::read_to_string;
use std::collections::HashMap;

// Moved to Python until I learn lifetimes :)

fn main() {
    let input = read_to_string("input.txt").unwrap();
    let mut root_map:HashMap<&str, Vec<&str>> = HashMap::new();
    input.split('\n').for_each(|l| {
        let mut vals = l.split('-');
        let a = vals.next().unwrap();
        let b = vals.next().unwrap();
        match root_map.get_mut(&a) {
            Some(v) => {v.push(b);},
            None => {root_map.insert(a, vec![b]);}
        };
        match root_map.get_mut(&b) {
            Some(v) => {v.push(a);},
            None => {root_map.insert(b, vec![a]);}
        };
    });
    println!("Root Map: {:?}", root_map);
    find_paths(&vec![], "start", &root_map);
    }


fn find_paths<'a> (prev_path: &'a Vec<&str>, node: &'a str, root_map: &'a HashMap<&str, Vec<&str>>) -> Vec<Vec<&'a str>> {
    let mut path = prev_path.clone();
    match node.parse::<char>(){
        Ok(c) => {
            if c.is_lowercase() && prev_path.contains(&node) {
                return vec![];
            }
        },
        _ => ()
    }
    if node == "start" && path != vec![] as Vec<&str> {
        return vec![];
    }
    path.push(node.as_ref());
    if node == "end" {
        let mut tmp: Vec<Vec<&str>> = vec![];
        tmp.push(Box::new(path));
        return tmp;
    }
    println!("to check: {:?}", root_map[node]);
    let mut children = vec![];
    for child in &root_map[node] {
        let val = find_paths(&path, child, &root_map);
        if !val.is_empty() {
            println!("{:?}", val);
            children.push(val);
        }
    }
    let out: Vec<Vec<&str>> = children.iter().flatten().map(|v| v.clone()).collect();
    println!("children: {:?}", out);
    out
}